idl/github.com/uber/cadence/cadence.thrift

```diff
@@ -262,4 +262,28 @@ service WorkflowService {
       3: shared.EntityNotExistsError entityNotExistError,
       4: shared.ServiceBusyError serviceBusyError,
     )
-}
\ No newline at end of file
+
+  /**
+  * RespondQueryTaskCompleted is called by application worker to complete a QueryTask (which is a DecisionTask for query)
+  * as a result of 'PollForDecisionTask' API call. Completing a QueryTask will unblock the client call to 'QueryWorkflow'
+  * API and return the query result to client as a response to 'QueryWorkflow' API call.
+  **/
+  void RespondQueryTaskCompleted(1: shared.RespondQueryTaskCompletedRequest completeRequest)
+    throws (
+      1: shared.BadRequestError badRequestError,
+      2: shared.InternalServiceError internalServiceError,
+      3: shared.EntityNotExistsError entityNotExistError,
+    )
+
+  /**
+  * QueryWorkflow returns query result for a specified workflow execution
+  **/
+  shared.QueryWorkflowResponse QueryWorkflow(1: shared.QueryWorkflowRequest queryRequest)
+	throws (
+	  1: shared.BadRequestError badRequestError,
+	  2: shared.InternalServiceError internalServiceError,
+	  3: shared.EntityNotExistsError entityNotExistError,
+	  4: shared.QueryFailedError queryFailedError,
```

Comment:
How is this different from internalServiceError?

Reply:
internalServiceError indicate something goes wrong on server. queryFailedError does not mean anything goes wrong on server. queryFailedError would happen when the worker cannot process the query. For example, when the query type is unknown, or when the query input is unexpected.